backend-comparison/src/burnbenchapp/auth.rs

```diff
@@ -0,0 +1,191 @@
+    let client = reqwest::blocking::Client::new();
+    // User-Agent is important otherwise GitHub rejects the request with a 403
+    // See: https://github.com/seanmonstar/reqwest/issues/918#issuecomment-632906966
+    let response = client
+        .get("https://api.github.com/user")
+        .header(reqwest::header::USER_AGENT, "burnbench")
+        .header(reqwest::header::ACCEPT, "application/vnd.github+json")
+        .header(reqwest::header::AUTHORIZATION, format!("Bearer {}", token))
+        .header(GITHUB_API_VERSION_HEADER, GITHUB_API_VERSION)
+        .send()?;
+    let response = response.json::<serde_json::Value>()?;
+    let username = response["login"].as_str().map(|s| s.to_string());
+    // Return an error if the login field is not found
+    username.ok_or_else(|| {
+        From::from(std::io::Error::new(
+            std::io::ErrorKind::NotFound,
+            "Username not found in the response",
+        ))
+    })
+}
+
+/// Save token in Burn cache directory and adjust file permissions
+pub(crate) fn save_token(token: &str) {
+    let path = get_auth_cache_file_path();
+    fs::create_dir_all(path.parent().expect("path should have a parent directory"))
+        .expect("directory should be created");
+    let mut file = File::create(&path).expect("file should be created");
+    write!(file, "{}", token).expect("token should be written to file");
+    // On unix systems we lower the permissions on the cache file to be readable
+    // just by the current user
+    #[cfg(unix)]
+    fs::set_permissions(&path, fs::Permissions::from_mode(0o600))
+        .expect("permissions should be set to 600");
+    println!("✅ Token saved at location: {}", path.to_str().unwrap());
+}
+
+/// Return the token saved in the cache file
+pub(crate) fn get_token_from_cache() -> Option<String> {
+    let path = get_auth_cache_file_path();
+    match fs::read_to_string(path) {
+        Ok(contents) => contents.lines().next().map(str::to_string),
+        _ => None,
+    }
+}
+
+#[cfg(test)]
+use serial_test::serial;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs;
+
+    fn cleanup_test_environment() {
+        let path = get_auth_cache_file_path();
+        if path.exists() {
+            fs::remove_file(&path).expect("should be able to delete the token file");
+        }
+        let parent_dir = path
+            .parent()
+            .expect("token file should have a parent directory");
+        if parent_dir.exists() {
+            fs::remove_dir_all(parent_dir).expect("should be able to delete the cache directory");
+        }
+    }
+
+    #[test]
+    #[serial]
+    fn test_save_token_when_file_does_not_exist() {
+        cleanup_test_environment();
+        let token = "unique_test_token";
+        // Ensure the file does not exist
+        let path = get_auth_cache_file_path();
+        if path.exists() {
+            fs::remove_file(&path).unwrap();
+        }
```

Comment:
Does the `cleanup_test_environement` do that?

Reply:
The test is about saving the token save when there is no cache file yet so I ensure that there is none at the beginning. I prefer to not assume that the state has been cleaned up correctly.